File: Java/207_course-schedule.java
// https://leetcode.com/problems/course-schedule/

import java.util.*;

class Solution {
public static boolean canFinish(int numCourses,int[][] prerequisites){
        if (prerequisites == null || prerequisites.length == 0) return true;
        
        //Will hold the graph
        HashMap<Integer,List<Integer>> adjacencyList = new HashMap<>();
        
		//pred[i] will hold the number of predecessors (prerequisites) there are for a given course
        int[] pred = new int[numCourses];
        
        //Build the graph and the predecessor counter
        for (int i = 0 ; i < prerequisites.length ; i = i + 1){
            int[] coursePair = prerequisites[i];
            
            int course = coursePair[0];
            int prereqCourse = coursePair[1];
            
            if (adjacencyList.get(course) == null) adjacencyList.put(course,new ArrayList<Integer>());
            
            adjacencyList.get(course).add(prereqCourse);
			
			//For each pair of courses you come upon [course,coursePrequisiteCourse], the number of prerequisites for the 
            pred[prereqCourse] += 1;
        }
        
        //If predecessor[i] = 0, that means that course does not have any prerequisites, add it to be processed next
        for (int i = 0 ; i < pred.length ; i++){
            
            if (pred[i] == 0){
                explore(i,adjacencyList,pred);
            }
            
        }
        
        //If there are any unexplored nodes in the graph, there is a cycle and you can't finish all courses
        for (int i = 0 ; i < pred.length ; i++){
            if (pred[i] != 0) return false;
        }
        
        return true;

    }

    //For the node that you pass in, you decrement all the predecessor counts on nodes that come after this one
    //If a given node during that has no more neighbors, you explore it
    
    //After you explore a given node, you remove it from the graph by setting its list of neighbors to null. Use this to keep track of nodes you've visited and to avoid going through old ones.
    public static void explore(int currentNode, HashMap<Integer,List<Integer>> graph, int[] predecessors){
        
        if (graph.get(currentNode) == null) return;
        
        for (Integer neighbor : graph.get(currentNode)){
            
            if (predecessors[neighbor] - 1 >= 0) predecessors[neighbor] = predecessors[neighbor] - 1;
            
            if (predecessors[neighbor] == 0) {
                explore(i,graph,predecessors);
            }
            
        }
        
        graph.put(currentNode,null);
    }

    public static void main(String args[]) {
        int numCourses =2;
        int prerequisites[][] = {{1,0},{0,1}};
        System.out.println(canFinish(numCourses, prerequisites));
    }
}

File: C++/207_course-schedule.cpp
// https://leetcode.com/problems/course-schedule/

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    bool canFinish(int numCourses, vector<vector<int>> &prerequisites)
    {
        /*
    To complete a course if you find  the same course again to complete then you are stuck inside a cycle.
    
    Cycle detection algorithm for Graph using DFS (Graph coloring algorithm)
    Each node has 3 states, 0 = not processed, 1 = processed, 2 = processing
    Getting a processing node while processing any other node is a cycle is the main logic here.
    */
        vector<int> visited(numCourses, 0);
        vector<vector<int>> adj(numCourses);
        for (auto course : prerequisites)
            adj[course[0]].push_back(course[1]); // this is adjacency matrix

        for (int i = 0; i < numCourses; i++)
        {
            if (visited[i] == 1)
                continue; // already processed

            if (isCycle(i, adj, visited))
                return false; // if cycle found then just return false
        }
        return true;
    }
    bool isCycle(int course, vector<vector<int>> &adj, vector<int> &visited)
    {
        if (visited[course] == 2)
            return true; // is processing but encountered so a cycle found

        visited[course] = 2; // mark it as processing
        for (auto couse_to_take : adj[course])
        {
            if (isCycle(couse_to_take, adj, visited))
                return true;
        }
        visited[course] = 1; // mark it as processed
        return false;
    }
};

int main()
{
    Solution a;
    int numCourses = 2;
    vector<vector<int>> prerequisites = {{1, 0}, {0, 1}};
    cout << a.canFinish(numCourses, prerequisites); //Output: false
    /*
    Explanation: There are a total of 2 courses to take. To take course 1 you should have finished course 0,
                 and to take course 0 you should also have finished course 1. So it is impossible.
    */
    return 0;
}